File: web/src/climbPerformance.test.ts
import { describe, expect, it } from "vitest";
import { climbPerformanceFor, gapForClimbLaps, paceForClimbSamples, samplesForClimbPerformance } from "./climbPerformance";
import type { ActivityClimb, ActivityLap, ActivitySample } from "./types";

const climb: ActivityClimb = {
  index: 0,
  difficulty: "Moderate",
  startSampleIndex: 0,
  endSampleIndex: 2,
  startDistanceM: 0,
  endDistanceM: 1000,
  distanceM: 1000,
  elevationGainM: 30,
  avgGradePct: 3,
  startElevationM: 100,
  endElevationM: 130
};

function samples(): ActivitySample[] {
  return [
    { index: 0, elapsedS: 0, distanceM: 0, speedMPS: 3.333333 },
    { index: 1, elapsedS: 150, distanceM: 500, speedMPS: 3.333333 },
    { index: 2, elapsedS: 300, distanceM: 1000, speedMPS: 3.333333 }
  ];
}

describe("climb performance", () => {
  it("computes climb pace from moving sample distance and time", () => {
    expect(paceForClimbSamples(samples(), climb)).toBeCloseTo(300);
  });

  it("uses bounded series samples when the activity detail omits raw samples", () => {
    const availableSamples = samplesForClimbPerformance(undefined, samples());
    expect(paceForClimbSamples(availableSamples, climb)).toBeCloseTo(300);
  });

  it("weights GAP by the lap distance that overlaps the climb", () => {
    const laps: ActivityLap[] = [
      { index: 0, elapsedTimeS: 300, movingTimeS: 300, distanceM: 500, avgGradeAdjustedPaceSPKM: 320 },
      { index: 1, elapsedTimeS: 300, movingTimeS: 300, distanceM: 500, avgGradeAdjustedPaceSPKM: 280 }
    ];
    expect(gapForClimbLaps(laps, climb)).toBeCloseTo(300);
  });

  it("omits performance when the source has no usable values", () => {
    const result = climbPerformanceFor([{ index: 0, distanceM: 0 }, { index: 2, distanceM: 1000 }], [], climb);
    expect(result.paceSPKM).toBeUndefined();
    expect(result.gapSPKM).toBeUndefined();
  });
});

File: internal/app/climb_performance_test.go
package app

import "testing"

func TestEnrichActivityClimbPerformanceUsesCompleteSamplesAndOverlappingLaps(t *testing.T) {
	climb := ActivityClimb{
		StartSampleIndex: 0,
		EndSampleIndex:   2,
		StartDistanceM:   0,
		EndDistanceM:     1000,
	}
	samples := []ActivitySample{
		{Index: 0, ElapsedS: intClimbPtr(0), DistanceM: floatClimbPtr(0), SpeedMPS: floatClimbPtr(10.0 / 3)},
		{Index: 1, ElapsedS: intClimbPtr(150), DistanceM: floatClimbPtr(500), SpeedMPS: floatClimbPtr(10.0 / 3)},
		{Index: 2, ElapsedS: intClimbPtr(300), DistanceM: floatClimbPtr(1000), SpeedMPS: floatClimbPtr(10.0 / 3)},
	}
	laps := []ActivityLap{
		{Index: 0, DistanceM: 500, AvgGradeAdjustedPaceSPKM: floatClimbPtr(320)},
		{Index: 1, DistanceM: 500, AvgGradeAdjustedPaceSPKM: floatClimbPtr(280)},
	}

	climbs := enrichActivityClimbPerformance([]ActivityClimb{climb}, samples, laps)
	if len(climbs) != 1 || climbs[0].PaceSPKM == nil || *climbs[0].PaceSPKM < 299.9 || *climbs[0].PaceSPKM > 300.1 {
		t.Fatalf("pace = %#v, want 300 s/km", climbs[0].PaceSPKM)
	}
	if climbs[0].GapSPKM == nil || *climbs[0].GapSPKM < 299.9 || *climbs[0].GapSPKM > 300.1 {
		t.Fatalf("GAP = %#v, want 300 s/km", climbs[0].GapSPKM)
	}
}

func TestEnrichActivityClimbPerformanceOmitsMissingValues(t *testing.T) {
	climb := ActivityClimb{StartSampleIndex: 0, EndSampleIndex: 1, StartDistanceM: 0, EndDistanceM: 500}
	climbs := enrichActivityClimbPerformance([]ActivityClimb{climb}, []ActivitySample{{Index: 0}, {Index: 1}}, nil)
	if climbs[0].PaceSPKM != nil || climbs[0].GapSPKM != nil {
		t.Fatalf("performance = %#v, want no values", climbs[0])
	}
}

func intClimbPtr(value int) *int {
	return &value
}

func floatClimbPtr(value float64) *float64 {
	return &value
}
